pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+				.ok_or(Error::<T>::LiquidityLimitNotStoredForAsset)?;
+
+			allowed_liquidity_range.update_amounts(amount_in, Zero::zero())?;
+			allowed_liquidity_range.check_limits()?;
+
+			<AllowedTradeVolumeLimitPerAsset<T>>::insert(asset_in, allowed_liquidity_range);
+		}
+
+		// liquidity out
+		// ignore Omnipool's hub asset
+		if asset_out != T::OmnipoolHubAsset::get() {
+			let mut allowed_liquidity_range = Pallet::<T>::allowed_trade_volume_limit_per_asset(asset_out)
+				.ok_or(Error::<T>::LiquidityLimitNotStoredForAsset)?;
+
+			allowed_liquidity_range.update_amounts(Zero::zero(), amount_out)?;
+			allowed_liquidity_range.check_limits()?;
+
+			<AllowedTradeVolumeLimitPerAsset<T>>::insert(asset_out, allowed_liquidity_range);
+		}
+
+		Ok(())
+	}
+
+	fn ensure_and_update_liquidity_limit(asset_id: T::AssetId, added_liquidity: T::Balance) -> DispatchResult {
+		if asset_id != T::OmnipoolHubAsset::get() && Pallet::<T>::liquidity_limit_per_asset(asset_id).is_some() {
+			let mut allowed_liquidity_limit = Pallet::<T>::allowed_liquidity_limit_per_asset(asset_id)
+				.ok_or(Error::<T>::LiquidityLimitNotStoredForAsset)?;
+
+			allowed_liquidity_limit.update_amount(added_liquidity)?;
+			allowed_liquidity_limit.check_limit()?;
+
+			<AllowedLiquidityAmountPerAsset<T>>::insert(asset_id, allowed_liquidity_limit);
+		}
+
+		Ok(())
+	}
+
+	pub fn validate_limit(limit: (u32, u32)) -> DispatchResult {
+		ensure!(
+			limit.0 <= MAX_LIMIT_VALUE && limit.1 <= MAX_LIMIT_VALUE,
+			Error::<T>::InvalidLimitValue
+		);
+		ensure!(!limit.0.is_zero() && !limit.1.is_zero(), Error::<T>::InvalidLimitValue);
+
+		Ok(())
+	}
+
+	pub fn calculate_limit(liquidity: T::Balance, limit: (u32, u32)) -> Result<T::Balance, DispatchError> {
+		let numerator = limit.0;
+		let denominator = limit.1;
+
+		// TODO: use u256
+		liquidity
+			.checked_mul(&T::Balance::from(numerator))
+			.ok_or(ArithmeticError::Overflow)?
+			.checked_div(&T::Balance::from(denominator))
+			.ok_or_else(|| ArithmeticError::DivisionByZero.into())
+	}
+}
+
+impl<T: Config> OnPoolStateChangeHandler<T::AssetId, T::Balance> for Pallet<T> {
+	fn after_pool_state_change(
+		asset_in: T::AssetId,
+		asset_in_reserve: T::Balance,
+		amount_in: T::Balance,
+		asset_out: T::AssetId,
+		asset_out_reserve: T::Balance,
+		amount_out: T::Balance,
+	) -> DispatchResult {
+		Pallet::<T>::calculate_and_store_trade_limit(asset_in, asset_in_reserve)?;
+		Pallet::<T>::calculate_and_store_trade_limit(asset_out, asset_out_reserve)?;
+		Pallet::<T>::ensure_and_update_trade_volume_limit(asset_in, amount_in, asset_out, amount_out)?;
+		Ok(())
+	}
+}
+
+impl<T: Config> OnLiquidityChangeHandler<T::AssetId, T::Balance> for Pallet<T> {
+	fn after_add_liquidity(
+		asset_id: T::AssetId,
+		initial_liquidity: T::Balance,
+		added_liquidity: T::Balance,
+	) -> DispatchResult {
+		Pallet::<T>::calculate_and_store_liquidity_limit(asset_id, initial_liquidity)?;
+		Pallet::<T>::ensure_and_update_liquidity_limit(asset_id, added_liquidity)?;
```

Comment:
we will probably want to ignore limits when the liquidity is added by the protocol (root origin)